src/coreclr/jit/importer.cpp

```diff
@@ -5303,70 +5303,83 @@ GenTree* Compiler::impOptimizeCastClassOrIsInst(GenTree* op1, CORINFO_RESOLVED_T
+            eeGetClassName(toClass));
 
-        if (castResult == TypeCompareState::Must)
-        {
-            // Cast will succeed, result is simply op1.
-            JITDUMP("Cast will succeed, optimizing to simply return input\n");
-            return op1;
-        }
-        else if (castResult == TypeCompareState::MustNot)
-        {
-            // See if we can sharpen exactness by looking for final classes
-            if (!isExact)
-            {
-                isExact = impIsClassExact(fromClass);
-            }
+    // Perhaps we know if the cast will succeed or fail.
+    TypeCompareState castResult = info.compCompHnd->compareTypesForCast(fromClass, toClass);
 
-            // Cast to exact type will fail. Handle case where we have
-            // an exact type (that is, fromClass is not a subtype)
-            // and we're not going to throw on failure.
-            if (isExact && !isCastClass)
-            {
-                JITDUMP("Cast will fail, optimizing to return null\n");
+    if (castResult == TypeCompareState::May)
+    {
+        JITDUMP("Result of cast unknown, must generate runtime test\n");
+        return nullptr;
+    }
 
-                // If the cast was fed by a box, we can remove that too.
-                if (op1->IsBoxedValue())
-                {
-                    JITDUMP("Also removing upstream box\n");
-                    gtTryRemoveBoxUpstreamEffects(op1);
-                }
+    if (castResult == TypeCompareState::Must)
+    {
+        // Even if op1 is more derived we know that it will still succeed.
+        JITDUMP("Cast will succeed, optimizing to simply return input\n");
+        return op1;
+    }
 
-                if (gtTreeHasSideEffects(op1, GTF_SIDE_EFFECT))
-                {
-                    impAppendTree(op1, CHECK_SPILL_ALL, impCurStmtDI);
-                }
-                return gtNewNull();
-            }
-            else if (isExact)
-            {
-                JITDUMP("Not optimizing failing castclass (yet)\n");
-            }
-            else
-            {
-                JITDUMP("Can't optimize since fromClass is inexact\n");
-            }
+    assert(castResult == TypeCompareState::MustNot);
+    if (isCastClass)
+    {
+        // TODO-CQ: Throw cast exception inline? Probably not important.
+        JITDUMP("Cannot optimize potentially failing castclass\n");
+        return nullptr;
+    }
+
+    // We know that an object of type fromClass cannot be cast to toClass, but
+    // if it is more derived it could still succeed a cast.
+    if (!isExact)
+    {
+        // We can still fail some type casts if they are between two types that
+        // belong to disjoint type hierarchies.
+        // TODO: compareTypesForCast(interfaceType, otherType) returning
+        // MustNot seems less useful than just having it return May. If we
+        // change this on EE side we can avoid these getClassAttribs calls.
```

Comment:
@jkotas do you think it makes sense to change this in `compareTypesForCast`? From what I understand, the current semantics of `compareTypesForCast` is "if I have an object of exact type A, can it be cast to type B" -- which is a meaningless question when `A` is an interface type. Alternatively, do you have any suggestions on better ways to check what we're after here?

Reply:
Hmm, looks like we fold `Type.IsAssignableFrom`/`Type.IsAssignableTo` using `compareTypesForCast` as well, so seems like we cannot really change its behavior.

Reply:
> do you think it makes sense to change this in compareTypesForCast?

Yes, I think this logic should be better done on the EE side.

> looks like we fold Type.IsAssignableFrom/Type.IsAssignableTo using compareTypesForCast as well, so seems like we cannot really change its behavior.

You can add a flag to the `compareTypesForCast` or add a new JIT/EE interface method.